app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
 
 import play.api.data.Form
 import play.api.data.Forms._
+import play.api.data.validation.{Constraint, Invalid, Valid, ValidationError}
 import play.api.mvc._
 import uk.gov.hmrc.play.frontend.controller.FrontendController
-import uk.gov.hmrc.selfservicetimetopay.controllerVariables
+import uk.gov.hmrc.selfservicetimetopay.connectors.CalculatorConnector
 import uk.gov.hmrc.selfservicetimetopay.controllerVariables._
 import uk.gov.hmrc.selfservicetimetopay.models._
+import uk.gov.hmrc.selfservicetimetopay.util.JacksonMapper
 import views.html.selfservicetimetopay.calculator._
 
 import scala.concurrent.Future
 
+//@TODO replace session with keysotre
 object CalculatorController extends FrontendController {
 
   private def createAmountDueForm: Form[CalculatorAmountDue] = {
     Form(mapping(
       "amount" -> bigDecimal,
-      "dueByDay" -> number(min = 1, max = 31),
+      "dueByYear" -> number(min = 2000, max = 2100),
       "dueByMonth" -> nonEmptyText,
-      "dueByYear" -> number(min = 2000, max = 2100)
+      "dueByDay" -> number(min = 1, max = 31)
     )(CalculatorAmountDue.apply)(CalculatorAmountDue.unapply))
   }
 
-  private def createPaymentTodayForm: Form[CalculatorPaymentToday] = {
+
+  private def createPaymentTodayForm(totalDue:BigDecimal): Form[CalculatorPaymentToday] = {
     Form(mapping(
-      "amount" -> optional(bigDecimal)
+      "amount" -> optional(bigDecimal
+        .verifying("ssttp.calculator.form.payment_today.amount.less-than-owed", _ < totalDue)
+        .verifying("ssttp.calculator.form.payment_today.amount.nonnegitive", _.compare(BigDecimal("0")) >= 0)
+    )
     )(CalculatorPaymentToday.apply)(CalculatorPaymentToday.unapply))
   }
 
-  private def createDurationForm(min: Integer, max: Integer): Form[CalculatorDuration] = {
+  private def createDurationForm(min: Int, max: Int): Form[CalculatorDuration] = {
+    def required: Constraint[Int] = Constraint[Int]("constraint.required") { o =>
+      if (o == null) Invalid(ValidationError("ssttp.calculator.form.duration.months.required")) else Valid
+    }
+    def greaterThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o < min) Invalid(ValidationError("ssttp.calculator.form.duration.months.greater-than", min)) else Valid
+    }
+    def lessThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o > max) Invalid(ValidationError("ssttp.calculator.form.duration.months.less-than", max)) else Valid
+    }
     Form(mapping(
-      "months" -> number(min = min, max = max)
+      "months" -> number.verifying(required, greaterThan, lessThan)
     )(CalculatorDuration.apply)(CalculatorDuration.unapply))
   }
 
+  private def getKeystoreData(implicit request:Request[AnyContent]) : (Boolean, Option[CalculatorAmountsDue], Option[CalculatorPaymentToday], Option[CalculatorDuration], Option[Seq[CalculatorPaymentSchedule]]) = {
```

Comment:
Payment Schedules are generated by an external service based on the rest of the data. The rest of the data is stored to allow the user to progress though each page and build up this information before getting the payment schedule. If data is missing then the user has to be be redirected to the necessary page to enter that data.

The flow is still subject to design change, and may change in the future, therefore it was better to get all the data and use another method to enforce the flow, but this is subject to to incremental improvement as the design and implementation matures. Keystore has yet to be implemented thus session is used in the interim to allow development to continue on the front end. Keeping in in a single method also isolates the rest of the controller methods from later changes/merge conflicts.